[SR-964] Warn if the implicit setter argument 'newValue' is unused

KingOfBrian opened the pull request:
This PR generates a warning if a setters implicit `newValue` variable is not used. If the variable is used, or if it is explicitly added to the setter (e.g. `set(newValue)`), no warning is generated.  This does not change willSet or didSet, since ignoring the implicit argument in observers is pretty common.

This turns out to create a lot of test churn! Before I run through all of the changes, I wanted to make sure that this is a desired warning, and that the behavior is correct. I think there are relatively few reasons to not use `newValue`, and disabling the warning is very trivial by adding `(newValue)`, or `(ignored)` to the setter. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-964](https://bugs.swift.org/browse/SR-964).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
Generate a warning if the implicit setter argument 'newValue' is not used.

tkremenek commented:
@swift-ci smoke test

DougGregor commented:
The test suite probably has a bunch of stub setters there only because we need the declaration to test something; that's probably not representative of real-world code. Still, I wonder if we should tune this heuristic a bit more. For example, we could warn only about setters that mention the property declaration itself but do not mention the `newValue` parameter (however it's spelled). That seems like it catches the intent of the warning better, because the user has likely typed the property name where she meant to type `newValue`.

DougGregor commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3516,6 +3516,10 @@ WARNING(pbd_never_used_stmtcond, none,
         (Identifier))
+WARNING(unused_implicit_setter_newvalue, none,
+        "implicit setter argument 'newValue' is never used; consider specifying"
+        " a variable name or using the argument `newValue`",
```
Per my other commentary, I don't think pinning it to the implicit parameter is quite right. I think it should always apply when the new value parameter is unused but the property is mentioned. Then, the suppression mechanism is the same as for other "unused" warnings:

    _ = newValue

rather than naming something you intentionally won't use.

KingOfBrian replied:
Great, much better, thanks @DougGregor! Thanks for the feedback too -- glad to get a re-direct before changing the 100 tests ;0

DougGregor commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1940,7 +1940,18 @@ class VarDeclUsageChecker : public ASTWalker {
+  bool isImplicitNewValueInSetter(VarDecl *VD) {
+    auto FD = dyn_cast<FuncDecl>(VD->getDeclContext());
+
+    return (VD->isImplicit() && VD->getName().str() == "newValue" &&
```
Rather than matching on the name string, check the appropriate parameter list of the enclosing function.

KingOfBrian replied:
Interesting, I've never distinguished between parameter lists, let me look into it.

Commit:
Change warning to only be reported if the parameter is unused but the getter is used.

KingOfBrian commented:
I updated the warning to complain only when the setter parameter is not used, and the getter is used. This behavior is much better and the change only generated 1 new warning in the test suite which I disabled with `_ = newValue`.

I removed some code that  added method parameters to the checker. It was only used to warn about var parameters that were not mutated. Since var parameter support was dropped in 3.0, this is just busy work.

KingOfBrian commented:
Hey @DougGregor -- just a ping on this, I updated the heuristic to match your recommendation.

KingOfBrian commented:
Hey @DougGregor -- Thought I'd try another ping on this.

milseman commented:
@swift-ci please smoke test

milseman commented:
@slavapestov can you steward this PR?

milseman commented:
@swift-ci please smoke test

milseman commented:
@swift-ci please smoke test

milseman commented:
@shahmishal swift-ci doesn't seem to be responsive. Is there an issue with the builders, the reporting mechanism, or this PR?

marcelofabri commented:
This happened on a PR that I created and the suggestion was to close it and open a new one (it worked).

milseman commented:
@DougGregor please review ;-)

shahmishal commented:
@swift-ci please smoke test

KingOfBrian commented:
Going to remake the PR to see if that appeases the build overlords.
